Todo 애플리케이션 테스트 코드 작성

BEOMKING opened the pull request:
## 요약
* 단위 테스트 코드 작성
* E2E 테스트 통과
* 테스트 커버리지 100% 달성

## 상세
현재 Todo 애플리케이션이 올바르게 동작하는지 확인할 수 있는 테스트 코드를 작성합니다.
컨트롤러, 서비스, 도메인 모델에 대한 테스트 코드를 작성해 테스트 커버리지를 100% 달성합니다.

Commit:
Test Path 생성 및 동작 확인

Commit:
작업 목록이 없을 때, 빈 배열을 리턴하는 테스트 작성

Commit:
작업 목록이 존재할 때, 조회 테스트

Commit:
Task 테스트를 위한 생성자와 Equals 정의

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -0,0 +1,46 @@
+        mockMvc.perform(get("/tasks"))
+                .andExpect(status().isOk());
+    }
+}
```
EOF에 빈칸을 추가해주세요!

BEOMKING replied:
엇 또 놓쳤군요,, 수정했습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @Test
+    @DisplayName("작업 목록이 주어지고 조회할 때, 작업 목록과 200 응답을 리턴한다.")
```
무엇을 테스트하는지 명확하게하기 위해서 BDD스타일을 접목해보시면 좋은 경험이 될 거에요!

BEOMKING replied:
종립님의 [글](https://johngrib.github.io/wiki/junit5-nested/#invalid-data-access-api-usage-exception-예외가-발생하는-테스트)을 참고해 Describe - Context - It 패턴의 BDD 스타일을 적용해보겠습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -0,0 +1,46 @@
+    @Test
+    @DisplayName("작업 목록이 주어지고 조회할 때, 작업 목록과 200 응답을 리턴한다.")
+    void GivenTasksAndWhenQueriedReturnsTasksAndOk() throws Exception {
+        given(taskService.getTasks()).willReturn(Arrays.asList(new Task(), new Task()));
```
id, title이 없어도 `할일` 이라고 할 수 있을까요?

BEOMKING replied:
주어진 상황을 잘못 작성한 것 같네요!

Commit:
Task 클래스의 메서드 테스트 코드 작성

getId, setId, getTitle, setTitle에 대한 테스트 코드를 작성했습니다.

BEOMKING committed:
EOF 추가

Commit:
TaskTest DCI 패턴 BDD 적용

Commit:
getTask 테스트 코드 작성

getTask의 주어진 식별자 유무 BDD 스타일의 테스트 코드를 작성했습니다.

Commit:
getTasks 테스트 코드 작성

Commit:
TaskService의 createTask 테스트 코드 작성

Commit:
TaskService의 updateTask 테스트 코드 작성

Commit:
updateTask 테스트 코드 추가

Commit:
EOF 추가

BEOMKING commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -110,5 +110,16 @@ void It_change_title_and_return() {
+            void It_throw_exception() {
+                assertThatThrownBy(() -> taskService.updateTask(givenId, givenChangeTask))
+                        .isExactlyInstanceOf(TaskNotFoundException.class);
+            }
```
현재 updateTask 메소드 내에서 getTask 메소드를 호출하기에 예외를 호출하고 있습니다.

이런 경우 getTask 메소드를 이미 테스트했으니 불필요한 코드일지 updateTask 메소드는 다른 결과를 호출할 가능성이 있을 수 있으니 필요한 코드인지 질문드립니다!

ssisksl77 replied:
한번 이렇게 생각해보아요! `getTask` 라는 구현과 상관없이 
`updateTask` 는 예외를 호출하는 것을 원하시나요? 아니면 다른 결과를 호출하기를 원하시나요?

BEOMKING replied:
`getTask` 구현에 따라 예외를 호출할 수도 하지 않을 수도 있습니다. 
`getTask`가 바뀌면 `updateTask`의 테스트 코드에도 영향을 미칠 수 있겠네요.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,125 @@
+    @DisplayName("createTask 메소드는")
+    class Describe_create_task {
+        @Nested
+        @DisplayName("작업이 주어졌을 때")
```
여기서 `작업` 은 아래에서 사용하는 `작업` 과는 다른 `식별자가 없는 작업` 일 것입니다.
그럼 여기서 `식별자가 없는 작업` 은 무엇을 말하는 걸까요?
자바 클래스의 관점에서 본다면 `id` 가 null 이면서 `title` 이 존재하는 것이겠죠!

그렇다면 `title` 만 있으면 작업이 되는 걸까요?

BEOMKING replied:
식별자와 제목을 모두 가지고 있어야 하나의 작업이 됩니다.

단지 입력을 `Task` 타입으로 받을 뿐 작업이라고 하긴 어렵겠네요.

[9c451f7](https://github.com/CodeSoom/spring-week3-assignment-1/pull/80/commits/9c451f76c47c16acbb91d9e9c6cfe9d53241b7a6)

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,125 @@
+            @Test
+            @DisplayName("식별자를 가진 작업을 생성하고 리턴한다")
+            void It_returns_task_with_id() {
+                assertThat(taskService.createTask(new Task(null, "BJP")))
```
`new Task(null, "BJP");` 에서 `null` 이 들어가지 않도록 개선해보아요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,125 @@
+    @DisplayName("updateTask 메소드는")
+    class Describe_update_task {
+        Long givenId = 1L;
+        Task givenTask = new Task(null, "변경전");
```
마찬가지로 null이 들어가지 않도록 변경해보아요! 어쩌면 구조를 많이 바꿔야할지도 몰라요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,79 @@
+            @Test
+            @DisplayName("작업에 제목을 할당한다")
+            void it_assignment_title_to_task() {
+                Task task = new Task();
+                task.setTitle(givenTitle);
```
setter를 되도록 사용하지 않도록해요!

그리고 가능하면 setter도 전부 없애보는 시도를 해보는거에요!

BEOMKING replied:
코드 구조를 바꿔 setter를 없애보았습니다.

Commit:
createTask 테스트 코드 상황, 리턴 설명 수정

Commit:
createTask 메소드 리팩터링

TaskService를 분리해서  TaskRepository를 만들었습니다.
컴파일 에러를 막기 위해 임시 주석처리를 했습니다.

Commit:
TaskService의 getTask 메소드 리팩터링

getTask 를 리팩터링했습니다. 이 과정에서 TaskRepository에 add 메서드가
추가되었으며 테스트 코드 또한 변경되었습니다.

Commit:
getTasks 메소드 리팩터링

getTasks 메소드를 리팩터링하면서 TaskRepository에 List 타입으로 작업
목록을 리턴해주는 메소드를 만들었고 테스트 코드 또한 변경했습니다.

Commit:
updateTask 메소드 리팩터링

updateTask를 리팩터링하면서 Task의 제목을 변경할 수 있는 changeTitle
메소드를 만들었습니다.
파라미터가 변경되면서 컨트롤러의 파라미터도 변경하였습니다.

Commit:
delete 메소드 리팩터링

delete 메소드를 리팩터링하면서 taskRepository에 remove 메소드를
만들었습니다.

Commit:
불필요한 필드, 테스트 코드 제거, change 테스트코드 추가

Repository로 분리하면서 사용하지 않는 필드 제거
setter를 제거하면서 setter 테스트 코드 제거
change 메소드에 대한 테스트 코드 추가

BEOMKING committed:
TaskController의 list 메소드 BDD 스타일로 변경

BEOMKING committed:
Merge branch 'BEOMKING' of https://github.com/BEOMKING/spring-week3-assignment-1 into BEOMKING

BEOMKING committed:
detail 메소드 테스트

Mockmvc와 Mockito를 이용해서 BDD 스타일의 detail 메소드를 작성했습니다.
작업을 가진 경우, 작업이 없는 경우 두 가지 테스트를 작성했습니다.

BEOMKING committed:
테스트 설명 잘못된 것 수정, 실패한 detail 테스트 코드 수정

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -1,23 +1,46 @@
     }
 
-    public void setTitle(String title) {
+    public Task changeTitle(String title) {
         this.title = title;
+        return this;
+    }
```
좋습니다. public 인 메소드는 주석을 추가해주시면 좋아요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,130 @@
+class TaskServiceTest {
+    TaskService taskService;
+    final Long givenId = 0L;
+    final String givenTitle = "BJP";
```
```suggestion
    private TaskService taskService;
    private final Long givenId = 0L;
    private final String givenTitle = "BJP";
```
nit: 접근제어자를 추가하는 것이 좋을 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,30 @@
+
+    public Task add(String title) {
+        Task task = new Task(sequence, title);
+        taskMap.put(sequence++, task);
```
어쩌면 ID를 만드는 기능을 분리할 수 있지 않을까!
라는 생각이 드네요!

경험삼아 ID생성 객체를 만들어 보아요!

BEOMKING replied:
IdGenerator를 생성해봤습니다.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,30 @@
+        return Optional.ofNullable(taskMap.get(id));
+    }
+
+    public List<Task> getAll() {
+        return new ArrayList<>(taskMap.values());
+    }
```
`taskMap.values()` 를 바로 리턴하도록 메소드 시그니처를 바꾸는 것에 대해서 어떻게 생각하시나요?

(이것에 정답은 없어요!)

BEOMKING replied:
레포지토리에서는 결과만 보내주고 서비스 레이어에서 자료구조를 결정할 수 있다는 장점이 있을 것 같아요. 
자료구조에 대한 요구사항 변경이 생겨도 모든 레이어가 아닌 서비스 레이어만 바꾸면 되겠네요!

BEOMKING committed:
Update app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java

Co-authored-by: 남영환 <ssisksl77@gmail.com>

BEOMKING committed:
mock을 초기화하는 reset 추가

Commit:
create 메소드 테스트 코드 추가

Commit:
deleteTask 메소드 시그니쳐 void 로 변경

void로 변경했기 때문에 리턴이 없는 경우 적용할 수 있는 테스트 코드로
변경 및 컨트롤러 테스트 코드 작성

Commit:
patch 테스트 request patch로 변경

BEOMKING committed:
작업의 식별자를 생성하는 기능을 가진 객체 생성

BEOMKING committed:
작업 목록 조회 시그니쳐 변경

List -> Collection으로 변경해서 서비스 레이어에서 자료 구조를 관리하므로 요구 사항 변화에 유연함을 가져갈 수 있습니다.

BEOMKING committed:
Task의 equals 메소드 테스트 추가

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -2,53 +2,40 @@
 import java.util.List;
 
 @Service
 public class TaskService {
```
Controller에서 사용해야 하는 기능들을 인터페이스로 분리해보시면 좋겠어요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/IdGenerator.java:
```diff
@@ -0,0 +1,20 @@
+ * 식별자를 생성하는 역할을 가지고 있습니다.
+ * 쓰레드 세이프한 식별자를 제공할 책임이 있습니다.
+ */
+public class IdGenerator {
```
이것도 인터페이스를 하나 만들어볼까요? IdGenerator라는 이름 자체는 꼭 Long만 리턴할 것 같진 않아요!

하나의 인터페이스를 만들고 그것에 대한 여러가지 구현체가 있으면 좋겠죠!

BEOMKING replied:
식별자로 Long이 올 수도 UUID 같은 것들이 올 수도 있을 것 같아요.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repository/TaskRepository.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.*;
+
+@Repository
+public class TaskRepository {
```
이것도 연습삼아 외부모듈에서 사용하는 기능들을 인터페이스에 분리하고 그것들을 구현하는 구조를 갖게 해보세요!

BEOMKING replied:
해당 구조로 바꾸고 Mock을 이용해 테스트 코드를 만들어보겠습니다.

BEOMKING committed:
TaskService 인터페이스 생성

TaskService 인터페이스를 생성하여 기존 TaskService는 Basic을 붙혀 구분하였습니다.

BEOMKING committed:
TaskRepository 인터페이스 생성

BEOMKING committed:
IdGenerator 인터페이스 생성

구현체로 Long, UUID Generator 생성
작업의 식별자를 여러 타입으로 만들 수 있습니다.

Commit:
BasicTaskService 클래스를 Mock을 이용해 테스트 코드 작성

BEOMKING commented on app/src/test/java/com/codesoom/assignment/application/BasicTaskServiceMockTest.java:
```diff
@@ -0,0 +1,75 @@
+            }
+        }
+
+        @Nested
+        @DisplayName("작업들이 없다면")
+        class Context_without_tasks {
+            @Test
+            @DisplayName("빈 작업을 리턴한다")
+            void It_returns_empty_tasks() {
+                HashMap<Long, Task> map = new HashMap<>();
+                given(taskRepository.getAll()).willReturn(map.values());
+
+                assertThat(taskService.getTasks()).isEmpty();
+
+                verify(taskRepository).getAll();
+            }
+        }
```
질문이 있습니다.

현재 TaskRepository 인터페이스의 getTasks() 메소드는 컬렉션을 반환하는 동작을 가지고 있습니다.
이 때, 
- TaskRepository의 A 구현체는 Task가 없다면 빈 컬렉션 리턴
- TaskRepository의 B 구현체는 Task가 없다면 기본 Task를 담아서 리턴
할 수도 있을 것 같아요. 

이럴 땐 구현체에 따라 TaskService의 테스트가 달라져야 할 것 같은데 
A, B 구현체를 사용하는 테스트를 따로 만들어야 할까요?

Mock을 사용하면 구현체가 달라졌을 때, 테스트가 깨지지 않을 것 같고
Mock을 사용하지 않고 구현한다면 구현체가 달라질 때마다 테스트도 변경해야 할 것 같습니다.

구현체에 따라 여러 테스트를 만들어야할까요?

ssisksl77 replied:
맞아요! Mock을 만들었을 때 구현체를 바꿔도 테스트가 깨지지 않는다는 것은 정말 위험한 문제에요!

그렇기 때문에 우리는 Mock 사용을 조심해야 하죠!

구현체가 여러개이며 구현체에 따라 테스트를 다르게 만든다면 훌륭합니다.

BEOMKING replied:
서비스의 구현체와 레포지토리의 구현체가 많아질수록 테스트 코드의 양이 많이 늘어나겠군요

BEOMKING replied:
요구사항이 변하게 되면 모든 연관된(컨트롤러, 서비스, 레포지토리)의 테스트 코드를 모두 변경해야 해서 유지보수할 것이 늘어나는 것 같기도 하고 테스트가 깨지면서 빨리 버그를 잡을 수 있어서 이게 맞는 것 같기도 합니다.
이번주 과제를 진행하면서 좀 혼란스러운 부분이있네요,, ㅎ

ssisksl77 replied:
구현체는 구현체마다 독립적으로 테스트를 할 것입니다.

요구사항이 변했을 때 모든 여러기능을 변경한다면 어떤 요구사항인지를 인지하는 것이 중요합니다!

또한 하나의 요구사항이 변할 때 다른 기능들의 테스트가 바뀐다면, 저는 다음처럼 생각할 것 같습니다.
1. 원래 변경되어야 할 부분이 1개의 모듈이 아닌 여러개의 모듈이다.
2. 현재 프로젝트의 코드구조가 1개의 모듈이 변경되어야 하는데 세부구현에 의존하는 다른 모듈이 존재한다.

BEOMKING committed:
BasicTaskServiceTest Mocking을 이용해 테스트 코드 작성

BEOMKING committed:
TaskController가 구현체가 아닌 인터페이스에 의존하도록 변경

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerWebTest.java:
```diff
@@ -20,16 +20,25 @@
     private MockMvc mockMvc;
     @MockBean
-    private BasicTaskService basicTaskService;
+    private TaskService taskService;
```
좋아요! 
이제 `@MockBean` 을 사용하지 않고!
`TaskService` 를 테스트용으로 하나 구현해보는 거에요!

BEOMKING replied:
`@MockMvc`를 사용하여 통합 테스트를 진행할 때, `@MockBean` 없이 제가 만든 구현체로 테스트 할 때 자동 주입이 되지 않아 `Failed to load ApplicationContext` 에러가 나왔습니다.
통합 테스트가 아닌 컨트롤러를 단위 테스트로 진행하는 것을 말씀하신게 맞을까요?
[eac98b3](https://github.com/CodeSoom/spring-week3-assignment-1/pull/80/commits/eac98b35f965bf7eccb29b03e04afa9cb6f2645e) 우선 `@MockMvc`를 사용하지 않는 단위 테스트 코드를 만들어보았습니다.

BEOMKING replied:
테스트용으로 구현하는 방법을 해보면서 느낀 점은
- 간단한 코드임에도 구현할 것이 많다.
- 실제 비즈니스 로직과 테스트용 로직의 결과를 실수로 다르게 할 수 있다.  
입니다.

ssisksl77 replied:
아주 좋습니다!

ssisksl77 replied:
우린 MockBean을 없애고 테스트를 진행해보았습니다. 이제는 `@SpringBootTest` 를 이용하여 통합테스트를 할 때 Bean에 살~짝! 테스트 빈을 올려보는 것을 연습해보아요!

어쩌면 이런 코드가 존재할 수 있겠어요!!

```java
@TestConfiguration
public class TestConfig {

    @Bean
    public TaskService taskService() {
        return new TaskService() {
            @Override
            public Collection<Task> getTasks() {
                return null;
            }

            @Override
            public Task getTask(Long id) {
                return null;
            }

            @Override
            public Task createTask(String title) {
                return null;
            }

            @Override
            public Task updateTask(Long id, String title) {
                return null;
            }

            @Override
            public void deleteTask(Long id) {

            }
        };
    }
}
```

BEOMKING committed:
TaskControllerTest 생성

TaskService를 테스트용 Fake 객체를 만들어서 테스트 진행했습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,214 @@
+            void It_returns_error_response_and_not_found() {
+                prepare();
+
+                assertThatThrownBy(() -> taskController.detail(givenId1))
+                        .isExactlyInstanceOf(TaskNotFoundException.class);
```
좋습니다. 하지만 이곳에서 테스트할 타겟은 `subject` 로 분리하면 더 가독성이 좋겠다는 생각이 드네요!

https://johngrib.github.io/wiki/junit5-nested/#subject-%EB%A9%94%EC%86%8C%EB%93%9C%EC%9D%98-%EC%82%AC%EC%9A%A9

BEOMKING committed:
테스트용 @Bean 등록후 사용 (매번 초기화하는 방법은 모르겠음)
